cmd/kubectl-direct_csi/drives_format.go

```diff
@@ -83,100 +81,68 @@ $ kubectl direct-csi drives format <drive_id_1> <drive_id_2>
+		if len(drives) == 0 && len(nodes) == 0 && len(accessTiers) == 0 {
+			return fmt.Errorf("atleast one of '%s', '%s' or '%s' should be specified",
+				utils.Bold("--all"),
+				utils.Bold("--drives"),
+				utils.Bold("--nodes"))
 		}
 	}
 
-	utils.Init()
 	directClient := utils.GetDirectCSIClient()
+	driveCh := getDrives(ctx, nodes, drives, accessTiers)
 
-	filterDrives := []directcsi.DirectCSIDrive{}
-	if len(args) > 0 {
-		for _, driveNameArg := range args {
-			driveName := strings.TrimSpace(driveNameArg)
-			drive, err := directClient.DirectCSIDrives().Get(ctx, driveName, metav1.GetOptions{})
-			if err != nil {
-				if !errors.IsNotFound(err) {
-					return err
-				}
-				klog.Errorf("No resource of %s found by the name %s", bold("DirectCSIDrive"), driveName)
-				continue
-			}
-			filterDrives = append(filterDrives, *drive)
-		}
-	} else {
-		driveList, err := directClient.DirectCSIDrives().List(ctx, metav1.ListOptions{})
-		if err != nil {
-			return err
-		}
-
-		if len(driveList.Items) == 0 {
-			klog.Errorf("No resource of %s found\n", bold("DirectCSIDrive"))
-			return fmt.Errorf("No resources found")
-		}
-
-		accessTierSet, aErr := getAccessTierSet(accessTiers)
-		if aErr != nil {
-			return aErr
+	wg := sync.WaitGroup{}
+	accessTierSet, aErr := getAccessTierSet(accessTiers)
+	if aErr != nil {
+		return aErr
+	}
+	for d := range driveCh {
+		if !d.MatchGlob(nodes, drives, status) {
+			continue
 		}
 
-		for _, d := range driveList.Items {
-			if d.MatchGlob(nodes, drives, status) {
-				if d.MatchAccessTier(accessTierSet) {
-					filterDrives = append(filterDrives, d)
-				}
-			}
+		if !d.MatchAccessTier(accessTierSet) {
+			continue
 		}
-	}
-
-	driveName := func(val string) string {
-		dr := strings.ReplaceAll(val, sys.DirectCSIDevRoot+"/", "")
-		dr = strings.ReplaceAll(dr, sys.HostDevRoot+"/", "")
-		return strings.ReplaceAll(dr, "-part-", "")
-	}
 
-	for _, d := range filterDrives {
 		if d.Status.DriveStatus == directcsi.DriveStatusUnavailable {
 			continue
 		}
 
+		path := canonicalNameFromPath(d.Status.Path)
+		nodeName := d.Status.NodeName
+		driveAddr := fmt.Sprintf("%s:/dev/%s", nodeName, path)
+
 		if d.Status.DriveStatus == directcsi.DriveStatusInUse {
-			driveAddr := fmt.Sprintf("%s:/dev/%s", d.Status.NodeName, driveName(d.Status.Path))
-			klog.Errorf("%s in use. Cannot be formatted", utils.Bold(driveAddr))
+			klog.Errorf("%s is in use. Cannot be formatted",
+				utils.Bold(driveAddr))
 			continue
 		}
 
-		if d.Status.DriveStatus == directcsi.DriveStatusReady {
-			driveAddr := fmt.Sprintf("%s:/dev/%s", d.Status.NodeName, driveName(d.Status.Path))
-			klog.Errorf("%s already owned and managed. Use %s to overwrite", utils.Bold(driveAddr), utils.Bold("--force"))
+		if d.Status.DriveStatus == directcsi.DriveStatusReady && !force {
+			klog.Errorf("%s is already owned and managed. Use %s to reformat",
+				utils.Bold(driveAddr), utils.Bold("--force"))
 			continue
```

Comment:
We have to make the drivestatus as 'available' to make the drive-controller accept the format request. Also, by moving the drive status from ready -> available, we ensure that this drive will not be participating in drive selection while creating volumes.